api/v1_projects.inc

```diff
@@ -576,3 +584,114 @@ function api_v1_projects_holdstates($method, $data, $query_params)
 {
     return Project::get_holdable_states();
 }
+
+// checkout a new page to proofread
+function api_v1_project_checkout($method, $data, $query_params)
+{
+    global $pguser;
+
+    $round = $data[":roundid"];
+    $project = $data[":projectid"];
+    validate_for_proofing($round, $project);
+
+    try {
+        $page_array = get_available_proof_page($project, $round, $pguser, $err);
+    } catch (ProjectPageException $exception) {
+        throw new BadRequest($exception->getMessage());
+    }
+    if (!$page_array) {
+        throw new NotFoundError($err, 1005);
+    }
+    [$pagename, $state] = $page_array;
+    // state will get set to temp
+    $proof_page = new ProofPage($project, $round, $pagename);
+    $proof_page->checkout();
+
+    return $proof_page->render_page_data();
+}
+
+function api_v1_project_page($method, $data, $query_params)
+{
+    global $PAGE_BADNESS_REASONS;
+
+    $round = $data[":roundid"];
+    $project = $data[":projectid"];
+    validate_for_proofing($round, $project);
+
+    $pagename = $data[":pagename"];
+    $proof_page = new ProofPage($project, $round, $pagename);
+
+    if ($method == "POST") {
+        // page could be in saved state
+        $proof_page->validate(true);
+        $proof_page->resume_saved_page();
+        return $proof_page->render_page_data();
+    }
+    // page should be in temp state
+    $proof_page->validate(false);
+    if ($method == "GET") {
+        $page_option = get_enumerated_param($query_params, "pageoption", "current", ["current", "original"]);
+        return $proof_page->render_page_text($page_option);
+    }
+    // must be "PUT"
+    $request_data = api_get_request_body();
+    $new_page_state = $request_data["new_page_state"] ?? null;
+    switch ($new_page_state) {
+    case "saved":
+        $text = $request_data["text"] ?? "";
+        $text_checker = new TextChecker($text, $project);
+        if (!$text_checker->all_good) {
+            return $text_checker->response();
+        }
+        $page_saved = $proof_page->attempt_save_as_done($text);
+        if (!$page_saved) {
+            throw new BadRequest(_("You have exceeded your daily page limit, page saved as in progress"), 1009);
+        }
+        break;
+    case "temp":
+        $text = $request_data["text"] ?? "";
+        $text_checker = new TextChecker($text, $project);
+        if (!$text_checker->all_good) {
+            return $text_checker->response();
+        }
+        $proof_page->save_as_in_progress($text);
+        break;
+    case "avail":
+        $proof_page->return_to_round();
+        break;
+    default:
+        // $new_page_state not valid
+        throw new BadRequest("invalid option for page state");
+        break;
+    }
+}
+
+function validate_for_proofing($round, $project)
+{
+    global $pguser;
+
+    if (!$round->user_access($pguser)->can_access) {
+        // The current user isn't allowed to work in the round
+        $err = sprintf(
```

Comment:
See about that strings returned by the API should not be translated / localized.